### PR TITLE
BUG: transpose inferring dtype for dt in object column

### LINKED ISSUE #51546: BUG: Transpose changes dtype from object to timedelta/datetime
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
df = DataFrame({"a": [pd.Timedelta("1 days"), pd.NaT, pd.Timedelta("2 days")]}, dtype=object)
df.T.dtypes


0    timedelta64[ns]
1     datetime64[ns]
2    timedelta64[ns]
dtype: object
```

### Issue Description

I'd expected this to return object, but this was changed in #26285

### Expected Behavior

The whatsnew looks like the fix was tailored to avoid raising an error, any other motivations?

### Installed Versions

<details>

Replace this line with the output of pd.show_versions()

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #51546 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

cc @jbrockmendel thoughts about doing it for 2.0 or 2.1?

### CODE REVIEW

File: pandas/tests/frame/methods/test_transpose.py [hunk 1]
```diff
@@ -63,7 +65,7 @@ def test_transpose_tzaware_2col_mixed_tz(self):
         df4 = DataFrame({"A": dti, "B": dti2})
         assert (df4.dtypes == [dti.dtype, dti2.dtype]).all()
         assert (df4.T.dtypes == object).all()
-        tm.assert_frame_equal(df4.T.T, df4)
+        tm.assert_frame_equal(df4.T.T, df4, check_dtype=False)
```
Comment (jbrockmendel) [inline pandas/tests/frame/methods/test_transpose.py:68]: can you cast the expected here to keep the test strict

File: pandas/tests/groupby/test_quantile.py [hunk 2]
```diff
@@ -464,7 +464,7 @@ def test_groupby_quantile_dt64tz_period():
 
     # Check that we match the group-by-group result
     exp = {i: df.iloc[i::5].quantile(0.5) for i in range(5)}
-    expected = DataFrame(exp).T
+    expected = DataFrame([x.tolist() for x in exp.values()], index=exp.keys())
```
Comment (jbrockmendel) [inline pandas/tests/groupby/test_quantile.py:467]: could do .infer_objects()?

**Comment by jbrockmendel** [comment]
> thoughts about doing it for 2.0 or 2.1?

no strong preference.  i think we're doing more still-merging-to-2.0 in the RC this time around than we usually do, not sure how much is too much

File: pandas/tests/frame/methods/test_transpose.py (hunk 1 above)
Comment (phofl) [reply pandas/tests/frame/methods/test_transpose.py:68]: done

**Comment by phofl** [comment]
Then lets do for 2.1, no strong preference either on my side

File: pandas/tests/groupby/test_quantile.py (hunk 2 above)
Comment (phofl) [reply pandas/tests/groupby/test_quantile.py]: Good idea

File: pandas/conftest.py [hunk 3]
```diff
@@ -1919,6 +1919,7 @@ def using_copy_on_write() -> bool:
     """
     Fixture to check if Copy-on-Write is enabled.
     """
+    pd.options.mode.copy_on_write = True
```
Comment (jbrockmendel) [inline pandas/conftest.py:1922]: is this intentional?

Comment (phofl) [reply pandas/conftest.py]: Nope thx, already removed. Snuck in

File: pandas/core/frame.py [hunk 4]
```diff
@@ -3597,7 +3601,9 @@ def transpose(self, *args, copy: bool = False) -> DataFrame:
             new_arr = self.values.T
             if copy:
                 new_arr = new_arr.copy()
-            result = self._constructor(new_arr, index=self.columns, columns=self.index)
+            result = self._constructor(
+                new_arr, index=self.columns, columns=self.index, dtype=new_arr.dtype
+            )
```
Comment (jbrockmendel) [inline pandas/core/frame.py:3606]: do we need to pass copy=False here (possibly just to be future-safe)?

**Review by jbrockmendel — APPROVED** [review]
LGTM

**Comment by jbrockmendel** [comment]
thanks @phofl

File: pandas/core/frame.py (hunk 4 above)
Comment (phofl) [reply pandas/core/frame.py:3606]: Yep, I’ll add it in the copy=True pr.

actually, it probably makes sense to go through all usages of constructor and pass copy=False for arrays. Will take a look how many there are

**Pull request merged**